### ISSUE TITLE
File transfers using direct connection

### ISSUE THREAD

**Issue by GoogleCodeExporter** [dequis/msn-pecan#35]
```
What steps will reproduce the problem?
1. attempt to send or receive a file via msn-pecan
2.
3.

What is the expected output? What do you see instead?
i would expect file transfer rates to reach a good speed ( 50kB/s or so?)
however i get terrible speeds

What version of the product are you using? On what operating system?
msn-pecan 0.0.11, pidgin 2.4.1, windows XP sp3 rc2

Please provide any additional information below.
it is definitely not the connection, since my house has a 20Mb/s connection
and i was receiving a file from my neighbour.


```

Original issue reported on code.google.com by `praveenm...@gmail.com` on 3 Apr 2008 at 3:33

Attachments:
- [direct file transfer.PNG](https://storage.googleapis.com/google-code-attachments/msn-pecan/issue-35/comment-0/direct file transfer.PNG)

**Comment by GoogleCodeExporter** [comment]
```
crap, i didnt blank out the emails on the attachment
```

Original comment by `praveenm...@gmail.com` on 3 Apr 2008 at 3:34

**Comment by GoogleCodeExporter** [comment]
```
That's because direct connection is disabled. Since all the I/O stuff was 
modified
there probably will be bugs.

I can try to enable it, but then I only tested with IP's in my same network.
```

Original comment by `felipe.contreras` on 3 Apr 2008 at 3:59
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
If you want, I could test it on my network. I have computers where one network 
card
has 2 IP addresses, one local and one public. Both computers have separate 
public IP
addresses as well.
```

Original comment by `Saturn2...@gmail.com` on 11 Apr 2008 at 8:14

**Comment by GoogleCodeExporter** [comment]
```
Saturn: Please add me to Google Talk so we can sync on this.
```

Original comment by `felipe.contreras` on 20 May 2008 at 6:26

**Comment by GoogleCodeExporter** [comment]
```
People, please vote on this. I'm not sure if to mark it as a requirement for 
0.1.0.

Use the star system.
```

Original comment by `felipe.contreras` on 23 May 2008 at 9:44

**Comment by GoogleCodeExporter** [comment]
```
This and offline messages are the two most important features to me personally. 
But
I'm not going to rush you :)
```

Original comment by `giraffeh...@gmail.com` on 23 May 2008 at 10:00

**Comment by GoogleCodeExporter** [comment]
```
i think its important. but like goo said, no rush. i do think offline messaging 
takes
priority over this though.
```

Original comment by `praveenm...@gmail.com` on 23 May 2008 at 10:04

**Comment by GoogleCodeExporter** [comment]
```
Cool. Still, can you use the star thing? It's at the top, left to the title.
```

Original comment by `felipe.contreras` on 23 May 2008 at 10:29

**Comment by GoogleCodeExporter** [comment]
```
Lack of direct transfers is the single reason why I don't use Pidgin 100% of the
time. Please, I hate using amsn.
```

Original comment by `matthaeu...@gmail.com` on 31 May 2008 at 1:01

**Comment by GoogleCodeExporter** [comment]
```
What is the status of this issue at the moment? It is in the specifications of
libmsn-pecan-0.0.14. Do you plan to implement upnp? Or do we need to open a 
specific
port for p2p transfers? 
```

Original comment by `marco.de...@point.nl` on 3 Jul 2008 at 9:51

**Comment by GoogleCodeExporter** [comment]
```
It's disabled right now. I need to update the code and even then it will work 
only in
some situations.

I'll work on this issue after oim support.
```

Original comment by `felipe.contreras` on 13 Jul 2008 at 11:42
- Added labels: **Priority-High**, **Milestone-0.1.0**
- Removed labels: **Priority-Medium**

**Comment by GoogleCodeExporter** [comment]
```
I think this should be one of the top priorities. This is the single reason A 
LOT of
my friends are still stuck with Live Messenger. Would it be possible to release 
a
beta-build or something so one could help test it for you?
```

Original comment by `justanot...@gmail.com` on 23 Jul 2008 at 1:20

**Comment by GoogleCodeExporter** [comment]
```
that would be a great idea
```

Original comment by `matthaeu...@gmail.com` on 27 Jul 2008 at 3:28

**Comment by GoogleCodeExporter** [comment]
```
I noticed that the OIM support is fixed. Does that mean that direct transfer 
support
should be here soon?
```

Original comment by `matthaeu...@gmail.com` on 21 Aug 2008 at 5:55

**Comment by GoogleCodeExporter** [comment]
```
It means that's the next thing I'll work on. I don't expect it to work on all 
cases
initially, it will be enabled case by case.
```

Original comment by `felipe.contreras` on 22 Aug 2008 at 12:44

**Comment by GoogleCodeExporter** [comment]
Original comment by `felipe.contreras` on 10 Sep 2008 at 10:08
- Added labels: **Type-Enhancement**
- Removed labels: **Type-Defect**

**Comment by GoogleCodeExporter** [comment]
Original comment by `devid...@gmail.com` on 14 Sep 2008 at 10:13
- Changed state: **Started**

**Comment by GoogleCodeExporter** [comment]
```
Hi!

What are the ports I need to open to get direct file transfer working?

Thanks for the amazing work!
```

Original comment by `lnogu...@gmail.com` on 8 Dec 2008 at 10:01

**Comment by GoogleCodeExporter** [comment]
```
lnogueir,
I would think that upnp would be in this build so you wouldn't have to worry 
about that
```

Original comment by `matthaeu...@gmail.com` on 11 Dec 2008 at 4:51

**Comment by GoogleCodeExporter** [comment]
```
Right now you can't act as a server, so the other end needs to have the ports 
open.

But yeah, ideally msn-pecan should use upnp. Is anyone willing to investigate 
out how
to ask for a port with upnp?
```

Original comment by `felipe.contreras` on 11 Dec 2008 at 6:17

**Comment by GoogleCodeExporter** [comment]
```
felipe, theres upnp.h in libpurple, but are you trying to not use libpurple too 
much 
these days?
```

Original comment by `eionrobb` on 11 Dec 2008 at 6:34

**Comment by GoogleCodeExporter** [comment]
```
eionrobb: exactly.

But I guess for starters that would do it.
```

Original comment by `felipe.contreras` on 11 Dec 2008 at 6:42

**Comment by GoogleCodeExporter** [comment]
```
Can someone write a summary on how the features of "direct-file-transfers" are 
atm?
Someone mentioned that i need to open up ports but now which ports, the same as 
the
msn-protocol?

Direct-file-transfer is a really great feature but needs some more work, 
counting on
you guys :P
```

Original comment by `sam...@gmail.com` on 17 Jan 2009 at 6:42

**Comment by GoogleCodeExporter** [comment]
```
use miniupnpc, the author made a patch for xchat and it's tiny. I assume the 
patch
for msn-pecan would also be quite small/easy. 
http://miniupnp.free.fr/xchat-upnp.html
```

Original comment by `weedy2...@gmail.com` on 17 Jan 2009 at 8:53

**Comment by GoogleCodeExporter** [comment]
```
I think this is the most important issue to fix.. i can test some new releases 
if you
want..
```

Original comment by `francesco.gatta@gmail.com` on 6 Feb 2009 at 7:55

**Comment by GoogleCodeExporter** [comment]
```
just compile it yourself from the git repo and try it out
```

Original comment by `matthaeu...@gmail.com` on 10 Feb 2009 at 5:24

**Comment by GoogleCodeExporter** [comment]
```
Hi, felipe any news as to when we will se an update to this issue?
Thanks

fede
```

Original comment by `fedeonda...@gmail.com` on 16 Mar 2009 at 12:20

**Comment by GoogleCodeExporter** [comment]
```
Hi, fast file transfers is the feature i miss most, Is there anything we can do 
to help ?
```

Original comment by `sw3t.s...@gmail.com` on 26 May 2009 at 12:40

**Comment by GoogleCodeExporter** [comment]
```
I'm marking this as critical for 0.1.0 since it's by far the most wanted 
feature.
I'll get back to work on this.
```

Original comment by `felipe.contreras` on 6 Jun 2009 at 1:12
- Changed title: **File transfers using direct connection**
- Added labels: **Priority-Critical**
- Removed labels: **Priority-High**

**Comment by GoogleCodeExporter** [comment]
```
the gupnp library, might be a good idea for use in msn-pecan. It is now the 
standard
for upnp in aMSN.

http://www.gupnp.org/
```

Original comment by `matthaeu...@gmail.com` on 12 Jul 2009 at 7:35

**Comment by GoogleCodeExporter** [comment]
```
Are there any updates for this feature?
```

Original comment by `baalbeh...@gmail.com` on 17 Sep 2009 at 3:26

**Comment by GoogleCodeExporter** [comment]
```
@matthaeus maybe libnice is a better choice?

@ballbehrit sorry, no updates yet... I've been busy but I'm confident at least 
some
basic support will be there on 0.1.0.
```

Original comment by `felipe.contreras` on 17 Sep 2009 at 9:48

**Comment by GoogleCodeExporter** [comment]
```
Ok, thanks for the answer. I'm pissed of 'cause I can't use pidgin because I 
need 
fast file transfer in MSN, and Windows Live Messenger drives me crazy. :)
```

Original comment by `baalbeh...@gmail.com` on 21 Sep 2009 at 12:38

**Comment by GoogleCodeExporter** [comment]
```
sorry to bother but, any news on this?
thanks
```

Original comment by `fedeonda...@gmail.com` on 30 Nov 2009 at 3:43

**Comment by GoogleCodeExporter** [comment]
```
@fedeondarts there's progress, now you can really try directconn with:
make DIRECTCONN=y

But it doesn't work properly. Please discuss in the mailing list if you want to 
help.
```

Original comment by `felipe.contreras` on 12 Dec 2009 at 10:10

**Comment by GoogleCodeExporter** [comment]
```
Somebody just released an alpha-grade patch for libpurple. Maybe this could 
help?

http://developer.pidgin.im/ticket/247#comment:49
```

Original comment by `matthaeu...@gmail.com` on 16 Dec 2009 at 1:29

**Comment by GoogleCodeExporter** [comment]
```
I've rewritten quite a bit of directconn and it seems to work fine for me, at 
least on 
internal networks :D

You can try on the latest git.

I just need to add a few tweaks so that external networks work too, and this 
bug will 
be fixed since it will work with most clients out there. Listening support and 
upnp 
should be tackled in another ticket.
```

Original comment by `felipe.contreras` on 29 Dec 2009 at 12:15

**Comment by GoogleCodeExporter** [comment]
```
Correction, only receiving works for now. I need to fix sending. Apparently the 
handshake is done the other way around =/
```

Original comment by `felipe.contreras` on 29 Dec 2009 at 1:39

**Comment by GoogleCodeExporter** [comment]
```
It turns out sending require many changes, but now it's working fine.

Only local networks work right now, the next step is to try external ips as 
well.

After that we could say we have basic support for direct connections.
```

Original comment by `felipe.contreras` on 25 Jan 2010 at 3:37

**Comment by GoogleCodeExporter** [comment]
```
Ok, I've pushed support for multiple ip addresses which means this bug is fixed 
:)

Commence the testing!

Further enhancements for direct connection should be handled in separate bug 
reports.
```

Original comment by `felipe.contreras` on 25 Jan 2010 at 9:51
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
```
Tested latest GIT on Pidgin 2.6.2.
Sending from Pidgin to WLM 8.5 (external) did not work. Transfer refused to 
start.
Receiving from WLM 8.5 (external) to Pidgin worked great.

On a side note, Pidgin sometimes randomly crashes while chatting.

Otherwise, great work!
```

Original comment by `andr...@gmail.com` on 26 Jan 2010 at 3:35

**Comment by GoogleCodeExporter** [comment]
```
@andrejx thanks for the testing. I'll investigate the issue while sending, I 
only 
tested on internal networks and it worked fine.
```

Original comment by `felipe.contreras` on 26 Jan 2010 at 11:42

**Comment by GoogleCodeExporter** [comment]
```
You've removed SHA1 hashing from my code (the original libpurple patch) so 
official
clients will refuse direct connection attempts with a mismatching 
nonce/nonce-hash.
If you're connecting to an official client (and not the other way around) then 
you
HAVE to have a correct nonce/nonce-hash pair.
```

Original comment by `kukker...@gmail.com` on 26 Jan 2010 at 4:22

**Comment by GoogleCodeExporter** [comment]
```
@kukkerman It seems you are assuming we are using your patch; we are not. I 
took a 
quick glance when you posted it, and that's it. If there's any similarity on 
the code, 
that's probably a coincidence.

Anyway, I don't understand what you are talking about. The nonce is created by 
the 
listening client, and then based on it both clients send the handshake 
messages. 
What's that hash you are talking about?
```

Original comment by `felipe.contreras` on 26 Jan 2010 at 8:21

**Comment by GoogleCodeExporter** [comment]
```
@kukkerman Ok, I looked at your patch and now I see what you mean. First of 
all, 
listening support is not yet implemented in msn-pecan, in those cases we would 
need 
to generate a nonce, but currently we don't have to: the other side does that.

However, you are doing extra steps while generating the nonce. The nonce string 
and 
binaries are straight-forward to convert from one to the other, when you 
generate the 
nonce you are using a sha1 checksum; that's unnecessary. It works because the 
data is 
random to begin with, so when you use a sha1 checksum you are just making it 
more 
random, you could use an md5, or rot13, or better: nothing.
```

Original comment by `felipe.contreras` on 26 Jan 2010 at 8:49

**Comment by GoogleCodeExporter** [comment]
```
I'm using SHA1 to generate the nonce-hash (hence the name) because during a 
direct
connect negotiation you have to send the nonce-hash first as a GUID. Then after 
the
TCP connection has been estabilished you have to send the nonce which you have
previously created the SHA1 hash from. If these two don't match the official 
client
drops the TCP connection. I'm not using SHA1 just for fun. :) I'm using it 
because
that's the way official clients work.

If listening support haven't implemented yet how will two msn-pecan clients 
connect
to each other? One party have to listen for incoming connections.
```

Original comment by `kukker...@gmail.com` on 26 Jan 2010 at 10:36

**Comment by GoogleCodeExporter** [comment]
```
@kukkerman You are the first one I see using "Hashed-Nonce" everybody else is 
using 
"Nonce". Also, I see the code where you generate a nonce string, but not code 
where 
you parse it.

msn-pecan clients cannot send files to each other yet.
```

Original comment by `felipe.contreras` on 27 Jan 2010 at 12:46

**Comment by GoogleCodeExporter** [comment]
```
The "Hashed-Nonce" is used by the official clients (Live Messenger and MSN 
Messenger
for example). At first I was surprised because I haven't found any information 
about
it. My suspicion is that they're used to increase security because when someone
connects to your opened socket you can check the nonce and the corresponding 
hash
acquired during direct connect negotiation. If they don't match then maybe 
someone
else is trying to "steal" files from you so you can drop the connection.
I think there's a pretty good chance that in the near future only the nonce-hash
method will be supported so it's better to prepare for that.

I don't parse the Nonce-Hash because my patch isn't complete. I've just 
released it
in its current state because I'm working on something else. You're right: the
Nonce-Hash should be parsed and checked against the acquired Nonce.
```

Original comment by `kukker...@gmail.com` on 27 Jan 2010 at 10:03

**Comment by GoogleCodeExporter** [comment]
```
@kukkerman I see, but I think before going the extra mile into what other 3rd 
party 
clients haven't even done yet; we need to provide basic support, and that's 
what I've 
tried to do in msn-pecan.

FTR. I'll release 0.1.0-rc3 so that people can start trying it, if there are no 
major 
issues, then we are ready for 0.1.0.

Afterwards, we'll need to implement listening support, then allow port 
configuration 
and then upnp. Only then we can start thinking about extra security stuff such 
as 
Hashed-Nonce. At least that's what I'll try to focus on... if somebody else 
sends a 
patch for any of that, it won't be rejected :)
```

Original comment by `felipe.contreras` on 27 Jan 2010 at 10:48

**Comment by GoogleCodeExporter** [comment]
```
I tried to send a file to a buddy using WLM2009 in the same LAN as me but he 
never
got the request transfer, using WinXP SP3 32 bit + pidgin 2.6.6 + 
pecan-0.1.0-rc4.

I'm attaching the log file in case it helps
```

Original comment by `tommaso....@gmail.com` on 2 Mar 2010 at 2:38

Attachments:
- [Copia di 02-03-2010_11.18.log](https://storage.googleapis.com/google-code-attachments/msn-pecan/issue-35/comment-50/Copia di 02-03-2010_11.18.log)

**Comment by GoogleCodeExporter** [comment]
```
I downloaded pecan-0.1.0 final and tested again.

Now requests reach the other buddy but the trasfer never start if direct 
connection
is selected in the account properties.

Setting direct connetcions to off the file transfer start, but if I stop it 
before
it's finished than Pidgin carshes. Please find attached rpt and log files.
```

Original comment by `tommaso....@gmail.com` on 4 Mar 2010 at 1:00

Attachments:
- [04-03-2010_13.34.log](https://storage.googleapis.com/google-code-attachments/msn-pecan/issue-35/comment-51/04-03-2010_13.34.log)
- [pidgin-crash-receiving-file.RPT](https://storage.googleapis.com/google-code-attachments/msn-pecan/issue-35/comment-51/pidgin-crash-receiving-file.RPT)

**Comment by GoogleCodeExporter** [comment]
```
So for people that have problem sending, there's issue #295 for that. I'm 
aiming to fix that for 0.2.

Is receiving working fine for everyone? I would like to mark this as verified.
```

Original comment by `felipe.contreras` on 16 May 2011 at 11:57
